goinstant/docker_test.go

```diff
@@ -0,0 +1,948 @@
+			wantErr:     false,
+			testInfo:    "mountPackage - should git clone custom package and copy it to the instant docker container",
+			runCommandMock: func(commandName string, suppressErrors []string, commandSlice ...string) (pathToPackage string, err error) {
+				pathToPackage = "./test"
+				if commandName == "git" {
+					return pathToPackage, nil
+				}
+				if commandName == "docker" && commandSlice[1] != pathToPackage {
+					t.Fatal("Path to package returned is incorrect")
+				}
+				return "", nil
+			},
+			unzipPackageMock: func(zipContent io.ReadCloser) (pathToPackage string, err error) {
+				return "", nil
+			},
+			untarPackageMock: func(tarContent io.ReadCloser) (pathToPackage string, err error) {
+				return "", nil
+			},
+		},
+		{
+			pathToPackage: "http://test.com/test1.zip",
+			mockServer: func() {
+				gock.New("http://test.com").
+					Get("/test1.zip").
+					Reply(200).
+					BodyString("Zip File Content")
+			},
+			errorString: "",
+			wantErr:     false,
+			testInfo:    "mountPackage - should unzip custom package and copy it to the instant docker container",
+			runCommandMock: func(commandName string, suppressErrors []string, commandSlice ...string) (pathToPackage string, err error) {
+				pathToPackage = "./test1"
+
+				if commandName == "docker" && commandSlice[1] != pathToPackage {
+					t.Fatal("Path to custom package returned is incorrect")
+				}
+				return "", nil
+			},
+			unzipPackageMock: func(zipContent io.ReadCloser) (pathToPackage string, err error) {
+				return "./test1", nil
+			},
+			untarPackageMock: func(tarContent io.ReadCloser) (pathToPackage string, err error) {
+				return "", nil
+			},
+		},
+		{
+			pathToPackage: "http://test.com/test2.tar.gz",
+			mockServer: func() {
+				gock.New("http://test.com").
+					Get("/test2.tar.gz").
+					Reply(200).
+					BodyString("Tar File Content")
+			},
+			errorString: "",
+			wantErr:     false,
+			testInfo:    "mountPackage - should untar custom package and copy it to the instant docker container",
+			runCommandMock: func(commandName string, suppressErrors []string, commandSlice ...string) (pathToPackage string, err error) {
+				pathToPackage = "./test2"
+
+				if commandName == "docker" && commandSlice[1] != pathToPackage {
+					t.Fatal("Path to custom package returned is incorrect")
+				}
+				return "", nil
+			},
+			unzipPackageMock: func(zipContent io.ReadCloser) (pathToPackage string, err error) {
+				return "", nil
+			},
+			untarPackageMock: func(tarContent io.ReadCloser) (pathToPackage string, err error) {
+				return "./test2", nil
+			},
+		},
+	}
+
+	for _, tt := range testCases {
+		runCommand = tt.runCommandMock
+		tt.mockServer()
+		unzipPackage = tt.unzipPackageMock
+		untarPackage = tt.untarPackageMock
+
+		t.Run(tt.testInfo, func(t *testing.T) {
+			err := mountCustomPackage(tt.pathToPackage)
+			if err == nil && tt.wantErr {
+				t.Fatal("Expected error - '" + tt.errorString + "' but got nil")
```

Comment:
If we're calling t.Fatal(), we're exiting the running of all test cases. t.Errorf(), on the other hand, will return an error for just that test case, but still run the other test cases. So, if we fail one test case, do we not want to run the other test cases?

Reply:
@Mark-Labuschagne. Thanks for the feedback man... Calling `t.Fatal` does not stop the other test cases from executing, it only stops the execution of that particular test case. `t.Errorf` marks the test case as failed, but does not stop the execution of that test case. Logic that should not be executed when the test fails (like the success messages we log out) will be executed when we use `t.Errorf`.

Reply:
Ahh cool, sorry for my confusion. What I think we should try then is do an else check that prints the test case passed. For instance
`
if (err != nil) != tt.wantErr {
				t.Errorf("unzipPackage() error = %v, wantErr %v", err, tt.wantErr)
				t.Log(tt.name, "failed!")
			} else {
				t.Log(tt.name, "passed!")
			}
`
This would print out which test case failed as well, making it a bit easier to spot the fail case...

Reply:
I'm checking now that this actually only works if we use log.Println() for printing... t.Log only prints information from the fail case